### PR TITLE
apiextensions: unify mono- and multi-versioned test helpers

### PR DESCRIPTION
**Pull request by sttts**
The mono-versioned helpers are a special case of the multi-versioned ones.

Fixes part of https://github.com/kubernetes/kubernetes/issues/64136.

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go [hunk 1]
```diff
@@ -42,30 +42,7 @@ import (
...
-		t.Fatal(err)
-	}
-	if e, a := definition.Spec.Group+"/"+definition.Spec.Version, createdTypeMeta.GetAPIVersion(); e != a {
-		t.Errorf("expected %v, got %v", e, a)
-	}
-	if e, a := definition.Spec.Names.Kind, createdTypeMeta.GetKind(); e != a {
-		t.Errorf("expected %v, got %v", e, a)
-	}
-	return createdInstance, nil
+	return instantiateVersionedCustomResource(t, instanceToCreate, client, definition, definition.Spec.Versions[0].Name)
```
Comment (mbohlool) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: should we call it `instantiateCustomResource` everywhere?

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go [hunk 2]
```diff
@@ -105,12 +82,7 @@ func NewNamespacedCustomResourceVersionedClient(ns string, client dynamic.Interf
 }
 
 func NewNamespacedCustomResourceClient(ns string, client dynamic.Interface, crd *apiextensionsv1beta1.CustomResourceDefinition) dynamic.ResourceInterface {
-	gvr := schema.GroupVersionResource{Group: crd.Spec.Group, Version: crd.Spec.Version, Resource: crd.Spec.Names.Plural}
-
-	if crd.Spec.Scope != apiextensionsv1beta1.ClusterScoped {
-		return client.Resource(gvr).Namespace(ns)
-	}
-	return client.Resource(gvr)
+	return NewNamespacedCustomResourceVersionedClient(ns, client, crd, crd.Spec.Versions[0].Name)
```
Comment (mbohlool) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: the same here as `NewNamespacedCustomResourceClient`

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go (hunk 1 above)
Comment (mbohlool) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: nit: (also previous one is nit) or we can call it the versioned but but I think we should just have one name for it.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go [hunk 3]
```diff
@@ -81,124 +81,163 @@ func TestClusterScopedCRUD(t *testing.T) {
 }
 
 func testSimpleCRUD(t *testing.T, ns string, noxuDefinition *apiextensionsv1beta1.CustomResourceDefinition, dynamicClient dynamic.Interface) {
```
Comment (mbohlool) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: just to make sure, this is the exact same code of `testSimpleVersionedCRUD`? I can skip reviewing it if it is :)

**Review by mbohlool — COMMENTED** [review]
just some nits. Looks good.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go [hunk 4]
```diff
@@ -81,124 +81,163 @@ func TestClusterScopedCRUD(t *testing.T) {
...
-		if e, a := watch.Added, watchEvent.Type; e != a {
-			t.Errorf("expected %v, got %v", e, a)
-			break
-		}
-		createdObjectMeta, err := meta.Accessor(watchEvent.Object)
-		if err != nil {
-			t.Fatal(err)
+	noxuResourceClients := map[string]dynamic.ResourceInterface{}
+	noxuWatchs := map[string]watch.Interface{}
+	disbaledVersions := map[string]bool{}
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: s/disbaled/disabled

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go [hunk 5]
```diff
@@ -105,12 +82,7 @@ func NewNamespacedCustomResourceVersionedClient(ns string, client dynamic.Interf
 }
 
 func NewNamespacedCustomResourceClient(ns string, client dynamic.Interface, crd *apiextensionsv1beta1.CustomResourceDefinition) dynamic.ResourceInterface {
```
Comment (nikhita) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: Move this to `resources.go` instead of here?

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: Same for `NewNamespacedCustomResourceVersionedClient`

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: had that before, but I don't want to add references to `Versions[0]` everywhere. That's becoming messy.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: Let's turn that around: rename `instantiateCustomResource` to `instantiateV1beta1CustomResource`

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: Leaving it as it is for now. Cannot come up with a less ugly variant.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: it is of course.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: fixed

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go]: it's still the same. maybe forgot to push? :)

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: don't want to promote these helpers. Instead I moved them into helpers.go and made them private. Similarly, some testserver helpers only used in k8s.io/apiextensions-apiserver I also moved into helpers.go.

**Comment by sttts** [comment]
@nikhita ptal. addressed the comments.

**Comment by nikhita** [comment]
lgtm from my side.

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go (hunk 5 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/registration_test.go]: :+1:

**Comment by nikhita** [comment]
/retest

so. many. flakes.

**Comment by mbohlool** [comment]
late /lgtm :)

**Pull request merged**